cmake/generic_toolchain.cmake

```diff
@@ -47,8 +47,8 @@ set(ZEPHYR_TOOLCHAIN_VARIANT ${ZEPHYR_TOOLCHAIN_VARIANT} CACHE STRING "Zephyr to
 assert(ZEPHYR_TOOLCHAIN_VARIANT "Zephyr toolchain variant invalid: please set the ZEPHYR_TOOLCHAIN_VARIANT-variable")
 
 # Pick host system's toolchain if we are targeting posix
-if((${ARCH} STREQUAL "posix") OR (${ARCH} STREQUAL "x86_64"))
-  set(ZEPHYR_TOOLCHAIN_VARIANT "host")
+if((${ARCH} STREQUAL "posix") OR (${ARCH} STREQUAL "x86_64") AND ("${ZEPHYR_TOOLCHAIN_VARIANT}" STREQUAL "zephyr"))
+    set(ZEPHYR_TOOLCHAIN_VARIANT "host")
```

Comment:
Can you explain this, it looks strange.

Reply:
This is to not change the current behaviour that is build native_posix using the host compiler (not the one provide in SDK). With this change if we explicitly set a specific TOOLCHAIN_VARIANT (in this case llvm) the build system will respect it.

Reply:
Regarding using `-Wl` I will check it again, but when I initially tried I got problem with clang because extra space that I couldn't solve, so I changed to use --for-linker.

Reply:
What if ZEPHYR_TOOLCHAIN_VARIANT=gnuarmemb ?

Do we want it to automatically choose host then? This was the old behaviour, but it changes with this patch.

Perhaps we should check for ZEPHYR_TOOLCHAIN_VARIANT != llvm instead. Or, now that we have multiple supported host toolchains we could require the user to choose. Before it made sense to choose for the user because there was only one choice, now we don't know what the user wants.

Reply:
> we could require the user to choose

I'd much rather not change the behavior in breaking ways. Meaning, today an user can be compiling for something else, and this piece of code allows them to suddenly compile native_posix without any environment change. Many users would be probably surprised and annoyed if suddenly this changes.

> now we don't know what the user wants.

We do guide users assuming they will compile the posix arch boards with the default host gcc in the documentation. So it would be logical to have it as the default.
(gcc is also the default compiler in Linux, and I'd guess we can assume most people will be using it)

Reply:
But this automatic switching, we want to support it for gnuarmemb as well don't we? Not just for ZEPHYR_TOOLCHAIN_VARIANT=zephyr.

Reply:
@SebastianBoe : Yes sure. I only meant to comment about the idea of not switching now by default to "host" due to having more host choices.
The gnuarmemb case would indeed be a regression

Reply:
I've updated it guys

Reply:
Add a comment above the if-statement and this is OK.

Something like;

> When targeting posix or 64bit x86, we make it easier for the user by automatically using the host GCC toolchain, even if the user has configured a different toolchain (except if the user has configured LLVM, in which case we assume the user does want the configured toolchain).